{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "416f06d0-2288-4df2-9bf5-35516b84b555",
   "metadata": {},
   "source": [
    "> Query whether the pseudo-sequence designed by NetMHCpan/MHCflurry makes sense.\n",
    "> In this notebook we fetch data from IPD database to get HLA protein sequences.\n",
    "> We implemented the pseudo-sequence defined by NetMHCpan, NetMHCpan2/4, MHCflurry2 respectively.\n",
    "\n",
    "## 1. Test if HLA alleles of the same group have identical pseudo-sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d5d44f2a-61dd-4990-b25e-ea3aa21e1cae",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"..\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "816ce087-c859-46a3-95b6-553b05cdc55a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   accession           name\n",
      "0   HLA00012  A*02:07:01:01\n",
      "1   HLA07326     A*02:07:02\n",
      "2   HLA10218     A*02:07:03\n",
      "3   HLA10811     A*02:07:04\n",
      "4   HLA10812     A*02:07:05\n",
      "5   HLA10905     A*02:07:06\n",
      "6   HLA12413     A*02:07:07\n",
      "7   HLA13174     A*02:07:08\n",
      "8   HLA17229     A*02:07:09\n",
      "9   HLA17578     A*02:07:10\n",
      "10  HLA17648     A*02:07:11\n",
      "11  HLA21755     A*02:07:12\n",
      "12  HLA23210     A*02:07:13\n",
      "13  HLA25642     A*02:07:14\n",
      "14  HLA25826  A*02:07:01:02\n",
      "15  HLA26783     A*02:07:15\n",
      "16  HLA27707     A*02:07:16\n",
      "17  HLA28005     A*02:07:17\n",
      "18  HLA29555     A*02:07:18\n",
      "19  HLA29578     A*02:07:19\n",
      "20  HLA30186     A*02:07:20\n",
      "21  HLA30286  A*02:07:01:03\n",
      "22  HLA32976  A*02:07:01:06\n",
      "23  HLA33233  A*02:07:01:05\n",
      "24  HLA33348     A*02:07:21\n",
      "25  HLA33357  A*02:07:01:04\n",
      "26  HLA35637     A*02:07:22\n",
      "27  HLA36701     A*02:07:23\n"
     ]
    }
   ],
   "source": [
    "from bioscript import IpdImgt\n",
    "import pandas\n",
    "import copy\n",
    "\n",
    "ipd_imgt = IpdImgt()\n",
    "\n",
    "hla_group = \"A*02:01\"\n",
    "\n",
    "hla_alleles = ipd_imgt.get_alleles(starts_with=hla_group)\n",
    "hla_iterator = ipd_imgt.iter(hla_alleles)\n",
    "hla_all = list()\n",
    "for item in hla_iterator:\n",
    "    hla_all.append(copy.deepcopy(item))\n",
    "\n",
    "df = pandas.DataFrame(hla_all)\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "39a6ed08-e92b-40c0-9db9-41aac4449aad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A*02:07:01:01 \t \n",
      "netmhcpan_ps: YFAMYGEKVAHTHVDTLYVRCHYYTWAVLAYTWY \t A*02:07P\n",
      "A*02:07:01:01 \t \n",
      "mhcflurry_ps: YFAMYGEKVAHTHVDTLYGVRCDHYYTWAVLAYTWYA \t A*02:07P\n"
     ]
    }
   ],
   "source": [
    "from bioscript import PseudoSequence as ps\n",
    "\n",
    "\n",
    "# Get a list of accession\n",
    "pseudo_seq_netmhcpan = list()\n",
    "pseudo_seq_netmhcpan_4 = list()\n",
    "pseudo_seq_netmhcpan_4_set = set()\n",
    "pseudo_seq_mhcflurry_2 = list()\n",
    "pseudo_seq_mhcflurry_2_set = set()\n",
    "for accession in list(df.accession):\n",
    "    single_allele = ipd_imgt.get_single_allele(accession)\n",
    "    if \"confirmation_status\" in single_allele and single_allele[\"confirmation_status\"][\"confirmed\"] is True and \"sequence\" in single_allele:\n",
    "        amino_seq = single_allele[\"sequence\"][\"protein\"]\n",
    "        pseudo_seq_netmhcpan.append(ps.sequence_to_pseudo_sequence(amino_seq, method=\"netmhcpan\"))\n",
    "        ps_netmhcpan_4 = ps.sequence_to_pseudo_sequence(amino_seq, method=\"netmhcpan_4\")\n",
    "        pseudo_seq_netmhcpan_4.append(ps_netmhcpan_4)\n",
    "        if ps_netmhcpan_4 not in pseudo_seq_netmhcpan_4_set:\n",
    "            print(single_allele[\"name\"], \"\\t\", \"\\nnetmhcpan_ps:\", ps_netmhcpan_4, \"\\t\",single_allele[\"wmda\"][\"P_group\"][\"name\"])\n",
    "            pseudo_seq_netmhcpan_4_set.add(ps_netmhcpan_4)\n",
    "        ps_mhcflurry_2 = ps.sequence_to_pseudo_sequence(amino_seq, method=\"mhcflurry_2\")\n",
    "        if ps_mhcflurry_2 not in pseudo_seq_mhcflurry_2_set:\n",
    "            print(single_allele[\"name\"], \"\\t\", \"\\nmhcflurry_ps:\", ps_mhcflurry_2, \"\\t\",single_allele[\"wmda\"][\"P_group\"][\"name\"])\n",
    "            pseudo_seq_mhcflurry_2_set.add(ps_mhcflurry_2)\n",
    "        pseudo_seq_mhcflurry_2.append(ps_mhcflurry_2)\n",
    "    else:\n",
    "        pseudo_seq_netmhcpan.append(\"\")\n",
    "        pseudo_seq_netmhcpan_4.append(\"\")\n",
    "        pseudo_seq_mhcflurry_2.append(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "89e775e7-0b7f-46bf-ae8e-80aba10ec4cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.insert(2, \"pseudo_seq_netmhcpan\", pseudo_seq_netmhcpan)\n",
    "df.insert(3, \"pseudo_seq_netmhcpan_4\", pseudo_seq_netmhcpan_4)\n",
    "df.insert(4, \"pseudo_seq_mhcflurry_2\", pseudo_seq_mhcflurry_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "53503b67-10dc-4412-828a-0bae46b1d12b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>name</th>\n",
       "      <th>pseudo_seq_netmhcpan</th>\n",
       "      <th>pseudo_seq_netmhcpan_4</th>\n",
       "      <th>pseudo_seq_mhcflurry_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HLA00012</td>\n",
       "      <td>A*02:07:01:01</td>\n",
       "      <td>RLARRSDSQMEAPIEGPWSTRASTDYLESTAATW</td>\n",
       "      <td>YFAMYGEKVAHTHVDTLYVRCHYYTWAVLAYTWY</td>\n",
       "      <td>YFAMYGEKVAHTHVDTLYGVRCDHYYTWAVLAYTWYA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HLA29578</td>\n",
       "      <td>A*02:07:19</td>\n",
       "      <td>RLARRSDSQMEAPIEGPWSTRASTDYLESTAATW</td>\n",
       "      <td>YFAMYGEKVAHTHVDTLYVRCHYYTWAVLAYTWY</td>\n",
       "      <td>YFAMYGEKVAHTHVDTLYGVRCDHYYTWAVLAYTWYA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   accession           name                pseudo_seq_netmhcpan  \\\n",
       "0   HLA00012  A*02:07:01:01  RLARRSDSQMEAPIEGPWSTRASTDYLESTAATW   \n",
       "19  HLA29578     A*02:07:19  RLARRSDSQMEAPIEGPWSTRASTDYLESTAATW   \n",
       "\n",
       "                pseudo_seq_netmhcpan_4                 pseudo_seq_mhcflurry_2  \n",
       "0   YFAMYGEKVAHTHVDTLYVRCHYYTWAVLAYTWY  YFAMYGEKVAHTHVDTLYGVRCDHYYTWAVLAYTWYA  \n",
       "19  YFAMYGEKVAHTHVDTLYVRCHYYTWAVLAYTWY  YFAMYGEKVAHTHVDTLYGVRCDHYYTWAVLAYTWYA  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filtered = df.replace(\"\", pandas.NA).dropna()\n",
    "df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d707be86-3f2c-4830-a9e8-b18c7b0bce5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pseudo_seq_netmhcpan_4\n",
      "YFAMYGEKVAHTHVDTLYVRCHYYTWAVLAYTWY    2\n",
      "dtype: int64\n",
      "pseudo_seq_netmhcpan\n",
      "RLARRSDSQMEAPIEGPWSTRASTDYLESTAATW    2\n",
      "dtype: int64\n",
      "pseudo_seq_mhcflurry_2\n",
      "YFAMYGEKVAHTHVDTLYGVRCDHYYTWAVLAYTWYA    2\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df_filtered.groupby(\"pseudo_seq_netmhcpan_4\").size())\n",
    "print(df_filtered.groupby(\"pseudo_seq_netmhcpan\").size())\n",
    "print(df_filtered.groupby(\"pseudo_seq_mhcflurry_2\").size())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "25683715-c6f6-4495-9b4a-b0eddb9d3874",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bioscript import IpdImgt\n",
    "from bioscript import PseudoSequence as ps\n",
    "import pandas\n",
    "import copy\n",
    "\n",
    "ipd_imgt = IpdImgt()\n",
    "\n",
    "def hla_to_pseudo_seq(hla_group):\n",
    "    hla_alleles = ipd_imgt.get_alleles(starts_with=hla_group)\n",
    "    hla_iterator = ipd_imgt.iter(hla_alleles)\n",
    "    hla_all = list()\n",
    "    for item in hla_iterator:\n",
    "        hla_all.append(copy.deepcopy(item))\n",
    "    \n",
    "    df = pandas.DataFrame(hla_all)\n",
    "    print(df)\n",
    "\n",
    "    # Get a list of accession\n",
    "    pseudo_seq_netmhcpan = list()\n",
    "    pseudo_seq_netmhcpan_4 = list()\n",
    "    pseudo_seq_netmhcpan_4_set = set()\n",
    "    pseudo_seq_mhcflurry_2 = list()\n",
    "    pseudo_seq_mhcflurry_2_set = set()\n",
    "    for accession in list(df.accession):\n",
    "        single_allele = ipd_imgt.get_single_allele(accession)\n",
    "        if \"confirmation_status\" in single_allele and single_allele[\"confirmation_status\"][\"confirmed\"] is True and \"sequence\" in single_allele:\n",
    "            if \"wmda\" in single_allele:\n",
    "                if \"P_group\" in single_allele[\"wmda\"]:\n",
    "                    amino_seq = single_allele[\"sequence\"][\"protein\"]\n",
    "                    pseudo_seq_netmhcpan.append(ps.sequence_to_pseudo_sequence(amino_seq, method=\"netmhcpan\"))\n",
    "                    ps_netmhcpan_4 = ps.sequence_to_pseudo_sequence(amino_seq, method=\"netmhcpan_4\")\n",
    "                    pseudo_seq_netmhcpan_4.append(ps_netmhcpan_4)\n",
    "                    if ps_netmhcpan_4 not in pseudo_seq_netmhcpan_4_set:\n",
    "                        print(single_allele[\"name\"], \"\\t\", \"\\nnetmhcpan_ps:\", ps_netmhcpan_4, \"\\t\",single_allele[\"wmda\"][\"P_group\"][\"name\"])\n",
    "                        pseudo_seq_netmhcpan_4_set.add(ps_netmhcpan_4)\n",
    "                    ps_mhcflurry_2 = ps.sequence_to_pseudo_sequence(amino_seq, method=\"mhcflurry_2\")\n",
    "                    if ps_mhcflurry_2 not in pseudo_seq_mhcflurry_2_set:\n",
    "                        print(single_allele[\"name\"], \"\\t\", \"\\nmhcflurry_ps:\", ps_mhcflurry_2, \"\\t\",single_allele[\"wmda\"][\"P_group\"][\"name\"])\n",
    "                        pseudo_seq_mhcflurry_2_set.add(ps_mhcflurry_2)\n",
    "                    pseudo_seq_mhcflurry_2.append(ps_mhcflurry_2)\n",
    "                else:\n",
    "                    pseudo_seq_netmhcpan.append(\"\")\n",
    "                    pseudo_seq_netmhcpan_4.append(\"\")\n",
    "                    pseudo_seq_mhcflurry_2.append(\"\")\n",
    "            else:\n",
    "                pseudo_seq_netmhcpan.append(\"\")\n",
    "                pseudo_seq_netmhcpan_4.append(\"\")\n",
    "                pseudo_seq_mhcflurry_2.append(\"\")\n",
    "        else:\n",
    "            pseudo_seq_netmhcpan.append(\"\")\n",
    "            pseudo_seq_netmhcpan_4.append(\"\")\n",
    "            pseudo_seq_mhcflurry_2.append(\"\")\n",
    "\n",
    "    df.insert(2, \"pseudo_seq_netmhcpan\", pseudo_seq_netmhcpan)\n",
    "    df.insert(3, \"pseudo_seq_netmhcpan_4\", pseudo_seq_netmhcpan_4)\n",
    "    df.insert(4, \"pseudo_seq_mhcflurry_2\", pseudo_seq_mhcflurry_2)\n",
    "\n",
    "    df_filtered = df.replace(\"\", pandas.NA).dropna()\n",
    "\n",
    "    print(df_filtered.groupby(\"pseudo_seq_netmhcpan_4\").size())\n",
    "    print(df_filtered.groupby(\"pseudo_seq_netmhcpan\").size())\n",
    "    print(df_filtered.groupby(\"pseudo_seq_mhcflurry_2\").size())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "20a6c9ed-af82-44e9-a9e9-a6673b98133a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   accession           name\n",
      "0   HLA00012  A*02:07:01:01\n",
      "1   HLA07326     A*02:07:02\n",
      "2   HLA10218     A*02:07:03\n",
      "3   HLA10811     A*02:07:04\n",
      "4   HLA10812     A*02:07:05\n",
      "5   HLA10905     A*02:07:06\n",
      "6   HLA12413     A*02:07:07\n",
      "7   HLA13174     A*02:07:08\n",
      "8   HLA17229     A*02:07:09\n",
      "9   HLA17578     A*02:07:10\n",
      "10  HLA17648     A*02:07:11\n",
      "11  HLA21755     A*02:07:12\n",
      "12  HLA23210     A*02:07:13\n",
      "13  HLA25642     A*02:07:14\n",
      "14  HLA25826  A*02:07:01:02\n",
      "15  HLA26783     A*02:07:15\n",
      "16  HLA27707     A*02:07:16\n",
      "17  HLA28005     A*02:07:17\n",
      "18  HLA29555     A*02:07:18\n",
      "19  HLA29578     A*02:07:19\n",
      "20  HLA30186     A*02:07:20\n",
      "21  HLA30286  A*02:07:01:03\n",
      "22  HLA32976  A*02:07:01:06\n",
      "23  HLA33233  A*02:07:01:05\n",
      "24  HLA33348     A*02:07:21\n",
      "25  HLA33357  A*02:07:01:04\n",
      "26  HLA35637     A*02:07:22\n",
      "27  HLA36701     A*02:07:23\n",
      "A*02:07:01:01 \t \n",
      "netmhcpan_ps: YFAMYGEKVAHTHVDTLYVRCHYYTWAVLAYTWY \t A*02:07P\n",
      "A*02:07:01:01 \t \n",
      "mhcflurry_ps: YFAMYGEKVAHTHVDTLYGVRCDHYYTWAVLAYTWYA \t A*02:07P\n",
      "pseudo_seq_netmhcpan_4\n",
      "YFAMYGEKVAHTHVDTLYVRCHYYTWAVLAYTWY    2\n",
      "dtype: int64\n",
      "pseudo_seq_netmhcpan\n",
      "RLARRSDSQMEAPIEGPWSTRASTDYLESTAATW    2\n",
      "dtype: int64\n",
      "pseudo_seq_mhcflurry_2\n",
      "YFAMYGEKVAHTHVDTLYGVRCDHYYTWAVLAYTWYA    2\n",
      "dtype: int64\n",
      "    accession           name\n",
      "0    HLA00097  A*31:01:02:01\n",
      "1    HLA02945     A*31:01:03\n",
      "2    HLA04466     A*31:01:04\n",
      "3    HLA04467     A*31:01:05\n",
      "4    HLA04548     A*31:01:06\n",
      "..        ...            ...\n",
      "104  HLA37257     A*31:01:52\n",
      "105  HLA37595     A*31:01:53\n",
      "106  HLA37778     A*31:01:54\n",
      "107  HLA38168  A*31:01:02:56\n",
      "108  HLA39006  A*31:01:02:57\n",
      "\n",
      "[109 rows x 2 columns]\n",
      "A*31:01:02:01 \t \n",
      "netmhcpan_ps: YTAMYQENVAHIDVDTLYIMYQDYTWAVLAYTWY \t A*31:01P\n",
      "A*31:01:02:01 \t \n",
      "mhcflurry_ps: YTAMYQENVAHIDVDTLYGIMYDQDYTWAVLAYTWYA \t A*31:01P\n",
      "HTTPSConnectionPool(host='www.ebi.ac.uk', port=443): Max retries exceeded with url: /cgi-bin/ipd/api/allele/HLA36802?project=HLA (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "pseudo_seq_netmhcpan_4\n",
      "YTAMYQENVAHIDVDTLYIMYQDYTWAVLAYTWY    28\n",
      "dtype: int64\n",
      "pseudo_seq_netmhcpan\n",
      "RLARRSDSQMEAPIERPWSIRASTDYLESTAATW    28\n",
      "dtype: int64\n",
      "pseudo_seq_mhcflurry_2\n",
      "YTAMYQENVAHIDVDTLYGIMYDQDYTWAVLAYTWYA    28\n",
      "dtype: int64\n",
      "   accession           name\n",
      "0   HLA00386  B*58:01:01:01\n",
      "1   HLA03248     B*58:01:02\n",
      "2   HLA04226     B*58:01:03\n",
      "3   HLA04388     B*58:01:04\n",
      "4   HLA05298     B*58:01:05\n",
      "..       ...            ...\n",
      "60  HLA36281  B*58:01:01:20\n",
      "61  HLA36517     B*58:01:44\n",
      "62  HLA36575     B*58:01:43\n",
      "63  HLA37364     B*58:01:45\n",
      "64  HLA38878     B*58:01:46\n",
      "\n",
      "[65 rows x 2 columns]\n",
      "B*58:01:01:01 \t \n",
      "netmhcpan_ps: YYATYGENMASTYENIAYIRYDSYTWAVLAYLWY \t B*58:01P\n",
      "B*58:01:01:01 \t \n",
      "mhcflurry_ps: YYATYGENMASTYENIAYGIRYDDSYTWAVLAYLWYA \t B*58:01P\n",
      "B*58:01:08 \t \n",
      "netmhcpan_ps: QVASYQSGSIIMYCDPDLGDIAQTLLGER \t B*58:01P\n",
      "B*58:01:08 \t \n",
      "mhcflurry_ps: QVASYQSGSIIMYCDPDLSGDINAQTLLGER \t B*58:01P\n",
      "pseudo_seq_netmhcpan_4\n",
      "QVASYQSGSIIMYCDPDLGDIAQTLLGER          1\n",
      "YYATYGENMASTYENIAYIRYDSYTWAVLAYLWY    16\n",
      "dtype: int64\n",
      "pseudo_seq_netmhcpan\n",
      "FTVEWETMKSAYRNLALYQMGQADQERARYLCLL     1\n",
      "RVARRSDSPTEAPIEGPWATRASIDYLESTAATW    16\n",
      "dtype: int64\n",
      "pseudo_seq_mhcflurry_2\n",
      "QVASYQSGSIIMYCDPDLSGDINAQTLLGER           1\n",
      "YYATYGENMASTYENIAYGIRYDDSYTWAVLAYLWYA    16\n",
      "dtype: int64\n",
      "    accession            name\n",
      "0    HLA00162   B*15:01:01:01\n",
      "1    HLA00163  B*15:01:01:02N\n",
      "2    HLA00164      B*15:01:02\n",
      "3    HLA00978      B*15:01:03\n",
      "4    HLA01373      B*15:01:04\n",
      "..        ...             ...\n",
      "143  HLA37571      B*15:01:81\n",
      "144  HLA37886      B*15:01:82\n",
      "145  HLA38696      B*15:01:83\n",
      "146  HLA39094      B*15:01:84\n",
      "147  HLA39217   B*15:01:01:66\n",
      "\n",
      "[148 rows x 2 columns]\n",
      "B*15:01:01:01 \t \n",
      "netmhcpan_ps: YYAMYREISTNTYESNLYLRYDSYTWAEWAYLWY \t B*15:01P\n",
      "B*15:01:01:01 \t \n",
      "mhcflurry_ps: YYAMYREISTNTYESNLYGLRYDDSYTWAEWAYLWYA \t B*15:01P\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Sequence contains invalid characters",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m hla_to_pseudo_seq(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mA*31:01\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      3\u001b[0m hla_to_pseudo_seq(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mB*58:01\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m----> 4\u001b[0m \u001b[43mhla_to_pseudo_seq\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mB*15:01\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      5\u001b[0m hla_to_pseudo_seq(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mC*03:02\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      6\u001b[0m hla_to_pseudo_seq(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mC*08:01\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "Cell \u001b[0;32mIn[2], line 28\u001b[0m, in \u001b[0;36mhla_to_pseudo_seq\u001b[0;34m(hla_group)\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mconfirmation_status\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m single_allele \u001b[38;5;129;01mand\u001b[39;00m single_allele[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mconfirmation_status\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mconfirmed\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msequence\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m single_allele:\n\u001b[1;32m     27\u001b[0m     amino_seq \u001b[38;5;241m=\u001b[39m single_allele[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msequence\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprotein\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m---> 28\u001b[0m     pseudo_seq_netmhcpan\u001b[38;5;241m.\u001b[39mappend(\u001b[43mps\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msequence_to_pseudo_sequence\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamino_seq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mnetmhcpan\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m)\n\u001b[1;32m     29\u001b[0m     ps_netmhcpan_4 \u001b[38;5;241m=\u001b[39m ps\u001b[38;5;241m.\u001b[39msequence_to_pseudo_sequence(amino_seq, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnetmhcpan_4\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     30\u001b[0m     pseudo_seq_netmhcpan_4\u001b[38;5;241m.\u001b[39mappend(ps_netmhcpan_4)\n",
      "File \u001b[0;32m~/git/BioScript/notebook/../bioscript/lib/pseudo_sequence/pseudo_sequence.py:21\u001b[0m, in \u001b[0;36mPseudoSequence.sequence_to_pseudo_sequence\u001b[0;34m(cls, seq_string, method)\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[38;5;129m@classmethod\u001b[39m\n\u001b[1;32m     18\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msequence_to_pseudo_sequence\u001b[39m(\u001b[38;5;28mcls\u001b[39m, seq_string, method):\n\u001b[1;32m     19\u001b[0m     \u001b[38;5;66;03m# Validate that the sequence only contains valid amino acid characters\u001b[39;00m\n\u001b[1;32m     20\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m re\u001b[38;5;241m.\u001b[39mfullmatch(\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m[ACDEFGHIKLMNPQRSTVWY]+\u001b[39m\u001b[38;5;124m'\u001b[39m, seq_string):\n\u001b[0;32m---> 21\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSequence contains invalid characters\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     23\u001b[0m     \u001b[38;5;66;03m# Extract the pseudo-sequence based on the given positions\u001b[39;00m\n\u001b[1;32m     24\u001b[0m     pseudo_sequence \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(seq_string[pos \u001b[38;5;241m-\u001b[39m \u001b[38;5;241m1\u001b[39m] \u001b[38;5;28;01mfor\u001b[39;00m pos \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mpositions[method] \u001b[38;5;28;01mif\u001b[39;00m pos \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(seq_string))\n",
      "\u001b[0;31mValueError\u001b[0m: Sequence contains invalid characters"
     ]
    }
   ],
   "source": [
    "hla_to_pseudo_seq(\"A*02:07\")\n",
    "hla_to_pseudo_seq(\"A*31:01\")\n",
    "hla_to_pseudo_seq(\"B*58:01\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5327fd0c-e51e-46c2-9813-6a604cf6a0ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    accession            name\n",
      "0    HLA00162   B*15:01:01:01\n",
      "1    HLA00163  B*15:01:01:02N\n",
      "2    HLA00164      B*15:01:02\n",
      "3    HLA00978      B*15:01:03\n",
      "4    HLA01373      B*15:01:04\n",
      "..        ...             ...\n",
      "143  HLA37571      B*15:01:81\n",
      "144  HLA37886      B*15:01:82\n",
      "145  HLA38696      B*15:01:83\n",
      "146  HLA39094      B*15:01:84\n",
      "147  HLA39217   B*15:01:01:66\n",
      "\n",
      "[148 rows x 2 columns]\n",
      "B*15:01:01:01 \t \n",
      "netmhcpan_ps: YYAMYREISTNTYESNLYLRYDSYTWAEWAYLWY \t B*15:01P\n",
      "B*15:01:01:01 \t \n",
      "mhcflurry_ps: YYAMYREISTNTYESNLYGLRYDDSYTWAEWAYLWYA \t B*15:01P\n",
      "B*15:01:09 \t \n",
      "netmhcpan_ps: QVANYQSGSTLMYCDPDLGDIAQTLLGER \t B*15:01P\n",
      "B*15:01:09 \t \n",
      "mhcflurry_ps: QVANYQSGSTLMYCDPDLSGDINAQTLLGER \t B*15:01P\n",
      "B*15:01:11 \t \n",
      "netmhcpan_ps: QVANYQSGSTLMYCDPDLGDIAQTLLGERDPHHD \t B*15:01P\n",
      "B*15:01:11 \t \n",
      "mhcflurry_ps: QVANYQSGSTLMYCDPDLSGDINAQTLLGERDPHHDQ \t B*15:01P\n",
      "HTTPSConnectionPool(host='www.ebi.ac.uk', port=443): Max retries exceeded with url: /cgi-bin/ipd/api/allele/HLA04216?project=HLA (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "pseudo_seq_netmhcpan_4\n",
      "QVANYQSGSTLMYCDPDLGDIAQTLLGER          7\n",
      "QVANYQSGSTLMYCDPDLGDIAQTLLGERDPHHD     1\n",
      "YYAMYREISTNTYESNLYLRYDSYTWAEWAYLWY    33\n",
      "dtype: int64\n",
      "pseudo_seq_netmhcpan\n",
      "FTVAWETSKNTYRSLLRYQMGQADQERARYLCLL     8\n",
      "RVARRSDSPMAAPIEGPWTTRASTDYLESTAATW    33\n",
      "dtype: int64\n",
      "pseudo_seq_mhcflurry_2\n",
      "QVANYQSGSTLMYCDPDLSGDINAQTLLGER           7\n",
      "QVANYQSGSTLMYCDPDLSGDINAQTLLGERDPHHDQ     1\n",
      "YYAMYREISTNTYESNLYGLRYDDSYTWAEWAYLWYA    33\n",
      "dtype: int64\n",
      "   accession           name\n",
      "0   HLA00410     C*03:02:01\n",
      "1   HLA01543  C*03:02:02:01\n",
      "2   HLA02800     C*03:02:03\n",
      "3   HLA04285     C*03:02:04\n",
      "4   HLA04287     C*03:02:05\n",
      "5   HLA04605     C*03:02:06\n",
      "6   HLA05731  C*03:02:02:02\n",
      "7   HLA06126     C*03:02:07\n",
      "8   HLA06420     C*03:02:08\n",
      "9   HLA07730     C*03:02:09\n",
      "10  HLA09569  C*03:02:02:03\n",
      "11  HLA09615     C*03:02:10\n",
      "12  HLA10103     C*03:02:11\n",
      "13  HLA10572     C*03:02:12\n",
      "14  HLA11874     C*03:02:13\n",
      "15  HLA13265     C*03:02:14\n",
      "16  HLA15212     C*03:02:15\n",
      "17  HLA17447     C*03:02:16\n",
      "18  HLA17571     C*03:02:17\n",
      "19  HLA17617  C*03:02:02:04\n",
      "20  HLA18968     C*03:02:18\n",
      "21  HLA19973     C*03:02:19\n",
      "22  HLA20143  C*03:02:02:05\n",
      "23  HLA20636     C*03:02:22\n",
      "24  HLA20738     C*03:02:21\n",
      "25  HLA21115     C*03:02:20\n",
      "26  HLA25161  C*03:02:02:06\n",
      "27  HLA25270  C*03:02:02:07\n",
      "28  HLA26261     C*03:02:23\n",
      "29  HLA27581  C*03:02:02:08\n",
      "30  HLA28139  C*03:02:02:09\n",
      "31  HLA28175     C*03:02:24\n",
      "32  HLA28296     C*03:02:25\n",
      "33  HLA29410  C*03:02:02:10\n",
      "34  HLA29442  C*03:02:02:11\n",
      "35  HLA29571  C*03:02:02:12\n",
      "36  HLA31129     C*03:02:26\n",
      "37  HLA33208  C*03:02:02:13\n",
      "38  HLA33316  C*03:02:02:14\n",
      "39  HLA33666  C*03:02:02:15\n",
      "40  HLA33708  C*03:02:02:16\n",
      "41  HLA34227     C*03:02:27\n",
      "42  HLA35550     C*03:02:28\n",
      "43  HLA36957     C*03:02:29\n",
      "44  HLA37516  C*03:02:02:17\n",
      "45  HLA37838     C*03:02:30\n",
      "46  HLA38729     C*03:02:31\n",
      "47  HLA39149     C*03:02:32\n",
      "C*03:02:01 \t \n",
      "netmhcpan_ps: YYAGYREKYRQTDVSNLYLRYDSYTWAELAYLWY \t C*03:02P\n",
      "C*03:02:01 \t \n",
      "mhcflurry_ps: YYAGYREKYRQTDVSNLYGLRYDDSYTWAELAYLWYA \t C*03:02P\n",
      "C*03:02:06 \t \n",
      "netmhcpan_ps: QVAQYQSGSILMYCDPDLGDIAQTLLGER \t C*03:02P\n",
      "C*03:02:06 \t \n",
      "mhcflurry_ps: QVAQYQSGSILMYCDPDLSGDINAQTLLGER \t C*03:02P\n",
      "C*03:02:10 \t \n",
      "netmhcpan_ps: QVAQYQSGSILMYCDPDLGDIAQTLLGEREHHHD \t C*03:02P\n",
      "C*03:02:10 \t \n",
      "mhcflurry_ps: QVAQYQSGSILMYCDPDLSGDINAQTLLGEREHHHDQ \t C*03:02P\n",
      "pseudo_seq_netmhcpan_4\n",
      "QVAQYQSGSILMYCDPDLGDIAQTLLGER         2\n",
      "QVAQYQSGSILMYCDPDLGDIAQTLLGEREHHHD    1\n",
      "YYAGYREKYRQTDVSNLYLRYDSYTWAELAYLWY    8\n",
      "dtype: int64\n",
      "pseudo_seq_netmhcpan\n",
      "FTVEWETYKQADRSLLRYQMGQADQERARYLCLL    3\n",
      "RLAHRSDSPGEAPVEGPWATRASIDYLESTAATW    8\n",
      "dtype: int64\n",
      "pseudo_seq_mhcflurry_2\n",
      "QVAQYQSGSILMYCDPDLSGDINAQTLLGER          2\n",
      "QVAQYQSGSILMYCDPDLSGDINAQTLLGEREHHHDQ    1\n",
      "YYAGYREKYRQTDVSNLYGLRYDDSYTWAELAYLWYA    8\n",
      "dtype: int64\n",
      "   accession           name\n",
      "0   HLA00445  C*08:01:01:01\n",
      "1   HLA01586     C*08:01:02\n",
      "2   HLA05786     C*08:01:03\n",
      "3   HLA06172     C*08:01:04\n",
      "4   HLA06644     C*08:01:05\n",
      "5   HLA08157     C*08:01:06\n",
      "6   HLA08158     C*08:01:07\n",
      "7   HLA08963     C*08:01:09\n",
      "8   HLA09834     C*08:01:10\n",
      "9   HLA10015     C*08:01:11\n",
      "10  HLA10017     C*08:01:12\n",
      "11  HLA10582     C*08:01:13\n",
      "12  HLA11095     C*08:01:14\n",
      "13  HLA11465     C*08:01:15\n",
      "14  HLA11571     C*08:01:16\n",
      "15  HLA11857     C*08:01:17\n",
      "16  HLA12333     C*08:01:18\n",
      "17  HLA13099     C*08:01:19\n",
      "18  HLA17225     C*08:01:20\n",
      "19  HLA17301  C*08:01:01:02\n",
      "20  HLA17463     C*08:01:21\n",
      "21  HLA19794     C*08:01:22\n",
      "22  HLA20343     C*08:01:24\n",
      "23  HLA20445     C*08:01:23\n",
      "24  HLA20782     C*08:01:25\n",
      "25  HLA20983     C*08:01:26\n",
      "26  HLA22259     C*08:01:27\n",
      "27  HLA22486     C*08:01:28\n",
      "28  HLA23140  C*08:01:01:04\n",
      "29  HLA23314  C*08:01:01:03\n",
      "30  HLA23758     C*08:01:29\n",
      "31  HLA24460  C*08:01:01:06\n",
      "32  HLA24807  C*08:01:01:05\n",
      "33  HLA24919  C*08:01:01:07\n",
      "34  HLA24930  C*08:01:01:08\n",
      "35  HLA26059  C*08:01:01:09\n",
      "36  HLA27113  C*08:01:01:10\n",
      "37  HLA27136  C*08:01:01:11\n",
      "38  HLA27595  C*08:01:01:12\n",
      "39  HLA27912  C*08:01:01:13\n",
      "40  HLA28885     C*08:01:30\n",
      "41  HLA29061     C*08:01:31\n",
      "42  HLA29208     C*08:01:32\n",
      "43  HLA29501  C*08:01:01:14\n",
      "44  HLA32683  C*08:01:01:16\n",
      "45  HLA33072  C*08:01:01:15\n",
      "46  HLA34250     C*08:01:08\n",
      "47  HLA34446  C*08:01:01:18\n",
      "48  HLA34612  C*08:01:01:19\n",
      "49  HLA34635  C*08:01:01:17\n",
      "50  HLA34941  C*08:01:01:20\n",
      "51  HLA35779     C*08:01:33\n",
      "52  HLA36505     C*08:01:34\n",
      "53  HLA36863     C*08:01:35\n",
      "54  HLA37384     C*08:01:36\n",
      "55  HLA38215  C*08:01:01:21\n",
      "56  HLA38303     C*08:01:37\n",
      "57  HLA38524  C*08:01:01:22\n",
      "C*08:01:01:01 \t \n",
      "netmhcpan_ps: YYAGYREKYRQTDVSNLYLRYNFYTWATLAYTWY \t C*08:01P\n",
      "C*08:01:01:01 \t \n",
      "mhcflurry_ps: YYAGYREKYRQTDVSNLYGLRYDNFYTWATLAYTWYA \t C*08:01P\n",
      "C*08:01:02 \t \n",
      "netmhcpan_ps: QVAQYQSGSTLMYCDPDLGDIAQTLLGKR \t C*08:01P\n",
      "C*08:01:02 \t \n",
      "mhcflurry_ps: QVAQYQSGSTLMYCDPDLFGDINAQTLLGKR \t C*08:01P\n",
      "C*08:01:10 \t \n",
      "netmhcpan_ps: QVAQYQSGSTLMYCDPDLGDIAQTLLGKREHHHD \t C*08:01P\n",
      "C*08:01:10 \t \n",
      "mhcflurry_ps: QVAQYQSGSTLMYCDPDLFGDINAQTLLGKREHHHDQ \t C*08:01P\n",
      "HTTPSConnectionPool(host='www.ebi.ac.uk', port=443): Max retries exceeded with url: /cgi-bin/ipd/api/allele/HLA11571?project=HLA (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "HTTPSConnectionPool(host='www.ebi.ac.uk', port=443): Max retries exceeded with url: /cgi-bin/ipd/api/allele/HLA11571?project=HLA (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "HTTPSConnectionPool(host='www.ebi.ac.uk', port=443): Max retries exceeded with url: /cgi-bin/ipd/api/allele/HLA11571?project=HLA (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "argument of type 'NoneType' is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m hla_to_pseudo_seq(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mB*15:01\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      2\u001b[0m hla_to_pseudo_seq(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mC*03:02\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m----> 3\u001b[0m \u001b[43mhla_to_pseudo_seq\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mC*08:01\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[4], line 26\u001b[0m, in \u001b[0;36mhla_to_pseudo_seq\u001b[0;34m(hla_group)\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m accession \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mlist\u001b[39m(df\u001b[38;5;241m.\u001b[39maccession):\n\u001b[1;32m     25\u001b[0m     single_allele \u001b[38;5;241m=\u001b[39m ipd_imgt\u001b[38;5;241m.\u001b[39mget_single_allele(accession)\n\u001b[0;32m---> 26\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mconfirmation_status\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43msingle_allele\u001b[49m \u001b[38;5;129;01mand\u001b[39;00m single_allele[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mconfirmation_status\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mconfirmed\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msequence\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m single_allele:\n\u001b[1;32m     27\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwmda\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m single_allele:\n\u001b[1;32m     28\u001b[0m             \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mP_group\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m single_allele[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwmda\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n",
      "\u001b[0;31mTypeError\u001b[0m: argument of type 'NoneType' is not iterable"
     ]
    }
   ],
   "source": [
    "hla_to_pseudo_seq(\"B*15:01\")\n",
    "hla_to_pseudo_seq(\"C*03:02\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6221a814-4bc6-408c-bce1-f5afe54c155e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   accession           name\n",
      "0   HLA00445  C*08:01:01:01\n",
      "1   HLA01586     C*08:01:02\n",
      "2   HLA05786     C*08:01:03\n",
      "3   HLA06172     C*08:01:04\n",
      "4   HLA06644     C*08:01:05\n",
      "5   HLA08157     C*08:01:06\n",
      "6   HLA08158     C*08:01:07\n",
      "7   HLA08963     C*08:01:09\n",
      "8   HLA09834     C*08:01:10\n",
      "9   HLA10015     C*08:01:11\n",
      "10  HLA10017     C*08:01:12\n",
      "11  HLA10582     C*08:01:13\n",
      "12  HLA11095     C*08:01:14\n",
      "13  HLA11465     C*08:01:15\n",
      "14  HLA11571     C*08:01:16\n",
      "15  HLA11857     C*08:01:17\n",
      "16  HLA12333     C*08:01:18\n",
      "17  HLA13099     C*08:01:19\n",
      "18  HLA17225     C*08:01:20\n",
      "19  HLA17301  C*08:01:01:02\n",
      "20  HLA17463     C*08:01:21\n",
      "21  HLA19794     C*08:01:22\n",
      "22  HLA20343     C*08:01:24\n",
      "23  HLA20445     C*08:01:23\n",
      "24  HLA20782     C*08:01:25\n",
      "25  HLA20983     C*08:01:26\n",
      "26  HLA22259     C*08:01:27\n",
      "27  HLA22486     C*08:01:28\n",
      "28  HLA23140  C*08:01:01:04\n",
      "29  HLA23314  C*08:01:01:03\n",
      "30  HLA23758     C*08:01:29\n",
      "31  HLA24460  C*08:01:01:06\n",
      "32  HLA24807  C*08:01:01:05\n",
      "33  HLA24919  C*08:01:01:07\n",
      "34  HLA24930  C*08:01:01:08\n",
      "35  HLA26059  C*08:01:01:09\n",
      "36  HLA27113  C*08:01:01:10\n",
      "37  HLA27136  C*08:01:01:11\n",
      "38  HLA27595  C*08:01:01:12\n",
      "39  HLA27912  C*08:01:01:13\n",
      "40  HLA28885     C*08:01:30\n",
      "41  HLA29061     C*08:01:31\n",
      "42  HLA29208     C*08:01:32\n",
      "43  HLA29501  C*08:01:01:14\n",
      "44  HLA32683  C*08:01:01:16\n",
      "45  HLA33072  C*08:01:01:15\n",
      "46  HLA34250     C*08:01:08\n",
      "47  HLA34446  C*08:01:01:18\n",
      "48  HLA34612  C*08:01:01:19\n",
      "49  HLA34635  C*08:01:01:17\n",
      "50  HLA34941  C*08:01:01:20\n",
      "51  HLA35779     C*08:01:33\n",
      "52  HLA36505     C*08:01:34\n",
      "53  HLA36863     C*08:01:35\n",
      "54  HLA37384     C*08:01:36\n",
      "55  HLA38215  C*08:01:01:21\n",
      "56  HLA38303     C*08:01:37\n",
      "57  HLA38524  C*08:01:01:22\n",
      "C*08:01:01:01 \t \n",
      "netmhcpan_ps: YYAGYREKYRQTDVSNLYLRYNFYTWATLAYTWY \t C*08:01P\n",
      "C*08:01:01:01 \t \n",
      "mhcflurry_ps: YYAGYREKYRQTDVSNLYGLRYDNFYTWATLAYTWYA \t C*08:01P\n",
      "C*08:01:02 \t \n",
      "netmhcpan_ps: QVAQYQSGSTLMYCDPDLGDIAQTLLGKR \t C*08:01P\n",
      "C*08:01:02 \t \n",
      "mhcflurry_ps: QVAQYQSGSTLMYCDPDLFGDINAQTLLGKR \t C*08:01P\n",
      "C*08:01:10 \t \n",
      "netmhcpan_ps: QVAQYQSGSTLMYCDPDLGDIAQTLLGKREHHHD \t C*08:01P\n",
      "C*08:01:10 \t \n",
      "mhcflurry_ps: QVAQYQSGSTLMYCDPDLFGDINAQTLLGKREHHHDQ \t C*08:01P\n",
      "pseudo_seq_netmhcpan_4\n",
      "QVAQYQSGSTLMYCDPDLGDIAQTLLGKR          2\n",
      "QVAQYQSGSTLMYCDPDLGDIAQTLLGKREHHHD     1\n",
      "YYAGYREKYRQTDVSNLYLRYNFYTWATLAYTWY    19\n",
      "dtype: int64\n",
      "pseudo_seq_netmhcpan\n",
      "FTVEWETYKQADRSLLRYQMGQADQERARYLCLL     3\n",
      "RLARQSDSPGEAPVEGPWATRASTDYLESTAATW    19\n",
      "dtype: int64\n",
      "pseudo_seq_mhcflurry_2\n",
      "QVAQYQSGSTLMYCDPDLFGDINAQTLLGKR           2\n",
      "QVAQYQSGSTLMYCDPDLFGDINAQTLLGKREHHHDQ     1\n",
      "YYAGYREKYRQTDVSNLYGLRYDNFYTWATLAYTWYA    19\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "hla_to_pseudo_seq(\"C*08:01\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "486a8a68-1bb0-477b-8e63-213834e52ebc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
